pkg/k8s/resources/resource.go

```diff
@@ -0,0 +1,16 @@
+package resources
+
+import "sigs.k8s.io/controller-runtime/pkg/client"
+
+// ResourceSpec allows generic Reconciler's implementation to get Hash value of the CR's Spec.
+// Hash is used to determine if spec has changed since the last status update or not.
+type ResourceSpec interface {
+	Hash() (string, error)
+}
+
+// Resource represents the custom resource object.
+type Resource interface {
+	client.Object
+	GetSpec() ResourceSpec
+	GetStatus() *ResourceStatus
```

Comment:
It is more idiomatic to name getters without the Get in Golang. Reference: [Effective Go](https://golang.org/doc/effective_go#Getters)

```suggestion
	Spec() ResourceSpec
	Status() *ResourceStatus
```

Reply:
Sorry, it collides with existing Spec and Status fields. Unfortunately, Go does not allow method and field names to be the same